vhost-device-spi/src/vhu_spi.rs

```diff
@@ -0,0 +1,1206 @@
+
+                writter
+                    .write(&reqs[i].rx_buf)
+                    .map_err(|_| Error::DescriptorWriteFailed)?;
+
+                len += rx_len;
+            } else {
+                writter
+                    .write_obj(in_hdr)
+                    .map_err(|_| Error::DescriptorWriteFailed)?;
+            }
+
+            // Write the transfer status
+            if vring.add_used(desc_chain.head_index(), len).is_err() {
+                warn!("Couldn't return used descriptors to the ring");
+            }
+        }
+
+        Ok(true)
+    }
+
+    /// Process the requests in the vring and dispatch replies
+    fn process_queue(&self, vring: &VringRwLock) -> Result<()> {
+        let Some(ref atomic_mem) = self.mem else {
+            return Err(Error::NoMemoryConfigured);
+        };
+        let requests: Vec<_> = vring
+            .get_mut()
+            .get_queue_mut()
+            .iter(atomic_mem.memory())
+            .map_err(|_| Error::DescriptorNotFound)?
+            .collect();
+
+        if self.process_requests(requests, vring)? {
+            // Send notification once all the requests are processed
+            vring
+                .signal_used_queue()
+                .map_err(|_| Error::NotificationFailed)?;
+        }
+
+        Ok(())
+    }
+}
+
+/// VhostUserBackendMut trait methods
+impl<D: 'static + SpiDevice + Sync + Send> VhostUserBackendMut for VhostUserSpiBackend<D> {
+    type Vring = VringRwLock;
+    type Bitmap = ();
+
+    fn num_queues(&self) -> usize {
+        NUM_QUEUES
+    }
+
+    fn max_queue_size(&self) -> usize {
+        QUEUE_SIZE
+    }
+
+    fn features(&self) -> u64 {
+        // this matches the current libvhost defaults except VHOST_F_LOG_ALL
+        1 << VIRTIO_F_VERSION_1
+            | 1 << VIRTIO_F_NOTIFY_ON_EMPTY
+            | 1 << VIRTIO_RING_F_INDIRECT_DESC
+            | 1 << VIRTIO_RING_F_EVENT_IDX
+            | VhostUserVirtioFeatures::PROTOCOL_FEATURES.bits()
+    }
+
+    fn protocol_features(&self) -> VhostUserProtocolFeatures {
+        VhostUserProtocolFeatures::MQ
+            | VhostUserProtocolFeatures::CONFIG
+            | VhostUserProtocolFeatures::REPLY_ACK
+    }
+
+    fn set_event_idx(&mut self, enabled: bool) {
+        dbg!(self.event_idx = enabled);
+    }
+
+    fn update_memory(&mut self, mem: GuestMemoryAtomic<GuestMemoryMmap>) -> IoResult<()> {
+        self.mem = Some(mem);
+        Ok(())
+    }
+
+    fn get_config(&self, offset: u32, size: u32) -> Vec<u8> {
+        // SAFETY: The layout of the structure is fixed and can be initialized by
+        // reading its content from byte array.
+        unsafe {
+            from_raw_parts(
+                self.spi_ctrl
+                    .config()
+                    .as_slice()
+                    .as_ptr()
+                    .offset(offset as isize) as *const _ as *const _,
+                size as usize,
+            )
+            .to_vec()
+        }
```

Comment:
Even if correct this looks weird. You can use `ByteValued::as_slice()` to convert the config to a `&[u8]` and copy its `offset...offset+size` range.

Reply:
As you suggest I update as follows and it works well:

```
fn get_config(&self, offset: u32, size: u32) -> Vec<u8> {
        self.spi_ctrl.config().as_slice()[(offset as usize)..(size as usize)].to_vec()
}
```

Is that fine? It does look clearer now.

Reply:
the slice needs to end at `offset + size`, but otherwise yes

Reply:
`self.spi_ctrl.config().as_slice()[(offset as usize)..][..size as usize)].to_vec()`

Reply:
Oh yes, the range is `offset ~ (offset + size)`,  thank you for correcting it.